### PR TITLE
storage/remote: pass RW 2.0 exemplars and metadata via AppendV2Options

### PR DESCRIPTION
**Pull request by krajorama**
Part of #17632. **Stacked on top of #19251** — base branch is `krajo/rw2-receiver-appenderv2`, so review only the top commit here.

## What

#19251 ported the RW 2.x receiver to `AppenderV2` but still appended exemplars and metadata through the standalone `AppendExemplar` / `UpdateMetadata` calls. This takes advantage of the combined `AppenderV2.Append`: for each received series, exemplars (sorted by timestamp, as the interface requires) and metadata are attached to the **first** sample/histogram append via `AppendV2Options`, and the remaining appends of the series carry empty options.

## Consequences

- **Exemplar stats/errors** now come back as `storage.AppendPartialError` from that first `Append`. `X-Prometheus-Remote-Write-Written-Exemplars` is derived from it (`appended = provided − errored`); out-of-order exemplars still map to `400`, other exemplar errors remain a no-op — preserving prior behaviour.
- **Metadata** is best-effort within `Append` and no longer reports an error, so the `samplesAppendedWithoutMetadata` increment is dropped from the RW 2.x path. The metric stays registered and is still used by the RW 1.0 path. No test asserts it.
- `remoteWriteAppenderV2.UpdateMetadata` is removed (no longer used).

## Edge case

Exemplars/metadata attach to the first sample/histogram, so if that first append is rejected (e.g. an out-of-order first sample) they are dropped with it; previously exemplars were appended in a separate pass. This only affects that specific case.

## Tests

The mock's V2 appender now records `opts.Exemplars`/`opts.Metadata` and returns partial errors, so the existing V2-message assertions hold. Added a case exercising **metadata stored via AppendV2Options** — metadata storage was previously not covered by the handler test.

```release-notes
NONE
```

### CODE REVIEW

File: storage/remote/write_handler.go [hunk 1]
```diff
@@ -384,20 +383,50 @@ func (h *writeHandler) appendV2(app *remoteWriteAppenderV2, req *writev2.Request
 			continue
 		}
 
-		allSamplesSoFar := rs.AllSamples()
+		// Parse the series' exemplars once. They (and the metadata) are attached to
+		// the first sample/histogram Append via AppendV2Options, taking advantage of
+		// the combined AppenderV2.Append. AppendV2Options.Exemplars must be sorted by
+		// timestamp.
+		var exemplars []exemplar.Exemplar
+		for _, ep := range ts.Exemplars {
```
Comment (bwplotka) [inline storage/remote/write_handler.go:386]: This is tricky given https://github.com/prometheus/prometheus/pull/19250#discussion_r3623444452 gap.

(Technically we should if Exemplar > 0 AND no sample and no histogram - we should likely write to standalone Exemplar for now. Then if it's related to sample/histogram we should either write to (A1) all(or (A2) first if we don't trust dedup logic) OR (B) change proto to have a new Exemplar field per sample so this logic is explicit.

File: storage/remote/write_handler.go [hunk 2]
```diff
@@ -384,20 +383,50 @@ func (h *writeHandler) appendV2(app *remoteWriteAppenderV2, req *writev2.Request
...
+			e, err := ep.ToExemplar(&b, req.Symbols)
+			if err != nil {
+				badRequestErrs = append(badRequestErrs, fmt.Errorf("parsing exemplar for series %v: %w", ls.String(), err))
+				continue
+			}
+			exemplars = append(exemplars, e)
+		}
+		slices.SortStableFunc(exemplars, func(a, b exemplar.Exemplar) int { return cmp.Compare(a.Ts, b.Ts) })
```
Comment (bwplotka) [inline storage/remote/write_handler.go:394]: We could also consider requiring this from protocol perspective

File: storage/remote/write_handler.go [hunk 3]
```diff
@@ -384,20 +383,50 @@ func (h *writeHandler) appendV2(app *remoteWriteAppenderV2, req *writev2.Request
...
+
+		opts := storage.AppendV2Options{Exemplars: exemplars}
+		// Only attach metadata (and thus persist it in the WAL) if the
+		// metadata-wal-records feature is enabled.
+		if h.appendMetadata {
+			opts.Metadata = m
+		}
+		// firstOpts returns opts for the first sample/histogram of the series and
+		// empty options for the rest, so exemplars and metadata are attached once.
+		firstOpts := func() storage.AppendV2Options {
```
Comment (bwplotka) [inline storage/remote/write_handler.go:404]: again, metadata should be safe to be attached again

Comment (bwplotka) [reply storage/remote/write_handler.go:404]: also in practice we shouldn't see penalty (we always send 1 sample)

**Review by bwplotka — COMMENTED** [review]
Interesting, I'd opt for a different logic

**Comment by krajorama** [comment]
favor #19457 over this
